rust/src/ffi/hashing.rs

```diff
@@ -214,9 +218,24 @@ pub unsafe extern "C" fn SCMd5HashBufferToHex(
 // Functions that are generic over Digest. For the most part the C bindings are
 // just wrappers around these.
 
-unsafe fn update<D: Digest>(digest: &mut D, bytes: *const u8, len: u32) {
+unsafe fn update<D: Digest>(digest: &mut D, bytes: *const u8, mut len: u32, bytes_hashed: *mut u32) -> u32 {
+    if !bytes_hashed.is_null() {
+        if g_hash_byte_limit > 0 {
+            let bytes_remaining = match g_hash_byte_limit > *bytes_hashed {
+                true => g_hash_byte_limit - *bytes_hashed,
+                false => 0,
+            };
+            if len > bytes_remaining {
+                len = bytes_remaining;
+            }
+        }
```

Comment:
I don't like the length logic being put into a low level hashing function which is not file-store specific. I think this accounting should be moved out of here.